### PR TITLE
Make an option to make addresses lower case

### PR DESCRIPTION
**Pull request by JasoonS**
TODO:
- [x] add tests
- [x] manually test some scenarios
- [x] think of more edge cases
- [x] get feedback on config file layout. For example Denham was keen to make this under a 'subgraph_compatibility' section in the config. I feel like it could also just be a general flag that someone uses so maybe it shouldn't be completely tied only to subgraphs?

## Summary by CodeRabbit

* **New Features**
  * Optional lowercase formatting for Ethereum addresses across generated projects; address helpers now validate/return lowercased addresses and clients forward checksum/lowercase choice.

* **Configuration**
  * New address_format option (checksum or lowercase) in configs and schema.
  * Generated runtime flags: lowercaseAddresses, shouldUseHypersyncClientDecoder, and allEventSignatures; templates updated to emit and thread them.

* **Validation**
  * Runtime checks prevent incompatible combinations of lowercaseAddresses and decoder settings.

* **Tests**
  * Added tests for address lowercasing and updated config/generation expectations.

### CODE REVIEW

File: codegenerator/cli/npm/envio/src/Address.res [hunk 1]
```diff
@@ -10,6 +10,28 @@ external unsafeFromString: string => t = "%identity"
 module Evm = {
   @module("viem")
   external fromStringOrThrow: string => t = "getAddress"
+
+  // NOTE: We could use a regex for this instead, not sure if it is faster/slower than viem's 'isAddress' function
+  //       `/^0x[a-fA-F0-9]{40}$/`
+  @module("viem") @private
```
Comment (DZakh) [inline codegenerator/cli/npm/envio/src/Address.res]:
```suggestion
  @module("viem")
```

@private doesn't exist in ReScript

File: codegenerator/cli/src/config_parsing/human_config.rs [hunk 2]
```diff
@@ -209,6 +209,11 @@ pub mod evm {
             description = "Makes handlers run twice to enable preload optimisations. Removes handlerWithLoader API, since it's not needed. (recommended, default: false)"
         )]
         pub preload_handlers: Option<bool>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        #[schemars(
+            description = "If true, Ethereum addresses are kept lowercase (no checksum) across the indexer (default: false)"
+        )]
+        pub lowercase_addresses: Option<bool>,
```
Comment (DZakh) [inline codegenerator/cli/src/config_parsing/human_config.rs]: I'd vote for `address_format: checksum | lowercase` field

File: codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res [hunk 3]
```diff
@@ -170,12 +172,24 @@ let make = (
...
+      )
+    } else {
+      HyperSyncClient.make(
+        ~url=endpointUrl,
+        ~apiToken,
+        ~maxNumRetries=Env.hyperSyncClientMaxRetries,
+        ~httpReqTimeoutMillis=Env.hyperSyncClientTimeoutMillis,
+        ~enableChecksumAddresses=true,
+      )
+    }
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res]: This can be simplified to `~enableChecksumAddresses=!lowercaseAddresses`

File: codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res [hunk 4]
```diff
@@ -188,7 +202,11 @@ let make = (
           ~msg="Failed to instantiate a decoder from hypersync client, please double check your ABI or try using 'event_decoder: viem' config option",
         )
       | decoder =>
-        decoder.enableChecksummedAddresses()
+        if lowercaseAddresses {
+          decoder.disableChecksummedAddresses()
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res:200]: Wow, I didn't know that we have the option

File: codegenerator/cli/npm/envio/src/Address.res (hunk 1 above)
Comment (JasoonS) [reply codegenerator/cli/npm/envio/src/Address.res]: Ah, funny. I actually added that. It wasn't from AI. I had a vague memory of something like this but you're right it might be from ReasonML :thinking:

We need to find a way to make sure this doesn't make Viem a required dependency for the users. I don't fully understand the solution that is used for the other function (`fromStringOrThrow`) but I assume I need to do the same here :thinking: - I thought making it private would maybe help.

File: codegenerator/cli/npm/envio/src/Address.res [hunk 5]
```diff
@@ -10,6 +10,28 @@ external unsafeFromString: string => t = "%identity"
...
+      )
+    }
+  }
+
+  let fromAddressLowercaseOrThrow = address =>
+    address->toString->fromStringLowercaseOrThrow
+
   // Reassign since the function might be used in the handler code
   // and we don't want to have a "viem" import there. It's needed to keep "viem" a dependency
   // of generated code instead of adding it to the indexer project dependencies.
```
Comment (JasoonS) [inline codegenerator/cli/npm/envio/src/Address.res:38]: Basically this comment is why I was trying to add `@private`

File: codegenerator/cli/npm/envio/src/Address.res (hunk 1 above)
Comment (JasoonS) [reply codegenerator/cli/npm/envio/src/Address.res]: Ok, it was this that I rememberred :sweat_smile: https://rescript-lang.org/docs/manual/v11.0.0/let-binding#private-let-bindings

Comment (DZakh) [reply codegenerator/cli/npm/envio/src/Address.res]: You can do `let fromStringOrThrow = fromStringOrThrow` below the line. This way it will be exported for ReScript users, but won't require to install viem

File: codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res (hunk 4 above)
Comment (JasoonS) [reply codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSyncSource.res:200]: I actually think it is the default, but doesn't hurt to explicitly set it :)

File: codegenerator/cli/npm/envio/src/Address.res [hunk 6]
```diff
@@ -10,6 +10,33 @@ external unsafeFromString: string => t = "%identity"
 module Evm = {
   @module("viem")
   external fromStringOrThrow: string => t = "getAddress"
```
[the remark this answers is not in the document]
Comment (JasoonS) [reply codegenerator/cli/npm/envio/src/Address.res:12]: Intentional overshadowing. Read comment.

File: codegenerator/cli/npm/envio/src/Address.res [hunk 7]
```diff
@@ -10,6 +10,29 @@ external unsafeFromString: string => t = "%identity"
...
+      unsafeFromString(string->Js.String2.toLowerCase)
+    } else {
+      Js.Exn.raiseError(
+        `Address "${string}" is invalid. Expected a 20-byte hex string starting with 0x.`,
+      )
+    }
+  }
+
+  let fromAddressLowercaseOrThrow = address =>
+    address->toString->fromStringLowercaseOrThrow
```
Comment (DZakh) [inline codegenerator/cli/npm/envio/src/Address.res]:
```suggestion
  let fromAddressLowercaseOrThrow = address =>
    address->toString->Js.String2.toLowerCase
```

Comment (JasoonS) [reply codegenerator/cli/npm/envio/src/Address.res]: I guess it should already be an address, makes sense :+1:
